File: 1792_maximum_average_pass_ratio.java
class Solution {
    public double maxAverageRatio(int[][] b, int n) {
        PriorityQueue<int[]> q = new PriorityQueue<>((a, c) -> Double.compare(
                (c[0] + 1.0) / (c[1] + 1.0) - c[0] * 1.0 / c[1],
                (a[0] + 1.0) / (a[1] + 1.0) - a[0] * 1.0 / a[1]));
        double w = 0;
        for (int[] k : b) {
            q.add(k);
            w += k[0] * 1.0 / k[1];
        }
        while (n-- > 0) {
            int[] h = q.poll();
            w += (h[0] + 1.0) / (h[1] + 1.0) - h[0] * 1.0 / h[1];
            h[0]++;
            h[1]++;
            q.add(h);
        }
        return w / b.length;
    }
}

File: 1792_maximum_average_pass_ratio.cpp
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& r, int o) {
        priority_queue<pair<double, int>> pq;
        double e = 0, h;
        int s = r.size();
        for (int i = 0; i < s; i++) {
            h = (double)r[i][0] / r[i][1];
            e += h / s;
            pq.push({(double)(r[i][0] + 1) / (r[i][1] + 1) - h, i});
        }
        while (o--) {
            auto c = pq.top();
            pq.pop();
            int d = c.second;
            e += c.first / s;
            r[d][0]++;
            r[d][1]++;
            h = (double)r[d][0] / r[d][1];
            pq.push({(double)(r[d][0] + 1) / (r[d][1] + 1) - h, d});
        }
        return e;
    }
};
